be/src/exec/pipeline/scan/connector_scan_operator.cpp

```diff
@@ -191,15 +192,35 @@ Status ConnectorChunkSource::_read_chunk(RuntimeState* state, vectorized::ChunkP
         return Status::EndOfFile("limit reach");
     }
 
-    do {
-        RETURN_IF_ERROR(_data_source->get_next(state, chunk));
-    } while ((*chunk)->num_rows() == 0);
+    while (_status.ok()) {
+        vectorized::ChunkPtr tmp;
+        _status = _data_source->get_next(state, &tmp);
+        if (_status.ok()) {
+            _ck_acc.push(std::move(tmp));
+            if (_ck_acc.output_chunk() != nullptr) {
+                break;
+            }
+        } else if (!_status.is_end_of_file()) {
+            return _status;
+        } else {
+            DCHECK(_status.is_end_of_file());
+        }
+    }
 
-    _rows_read += (*chunk)->num_rows();
+    DCHECK(_status.ok() || _status.is_end_of_file());
     _scan_rows_num = _data_source->raw_rows_read();
     _scan_bytes = _data_source->num_bytes_read();
-
-    return Status::OK();
+    if (_ck_acc.output_chunk() != nullptr || _ck_acc.staging_chunk() != nullptr) {
+        if (_ck_acc.output_chunk() != nullptr) {
+            *chunk = std::move(_ck_acc.output_chunk());
+        } else {
+            *chunk = std::move(_ck_acc.staging_chunk());
+        }
+        _rows_read += (*chunk)->num_rows();
+        return Status::OK();
+    }
```

Comment:
How about also extracting `pull_chunk()`, `is_finished()` and `set_finished()`, etc. to `ChunkPipelineAccumulator` like follows? Then maybe we could hide the logic about `output_chunk` and `staging_chunk`, and needn't write some duplicated code.

```C++
while (_status.ok() && _ck_acc.need_input()) {
    _status = _data_source->get_next(state, &tmp);
    if (_status.ok()) {
        _ck_acc.push(std::move(tmp));
    } else if (!_status.is_end_of_file()) {
        return _status;
    } else {
        _ck_acc.set_finishing();
    }
}

// ...
if (_ck_acc.has_output()) {
    *chunk = _ck_acc.pull_chunk();
    return Status::OK();
}

_ck_acc.reset();
return Status::EndOfFile("");

```

Reply:
Thanks. It's much more clear